File: legacylift/server/core/target_languages.py
"""
core/target_languages.py - Canonical backend target-language metadata.

This module is intentionally small and dependency-free so generation, static
validation, test generation, and registry code can agree on target ids without
importing the heavier Layer 0.5 registry.
"""

from __future__ import annotations

import re
from dataclasses import dataclass


@dataclass(frozen=True)
class TargetLanguageInfo:
    id: str
    label: str
    language: str
    version: str
    extension: str
    comment_prefix: str
    code_fence: str
    test_framework: str
    numeric_policy: str
    date_policy: str
    style_guide: str
    type_system: str
    async_model: str
    recommended_libraries: tuple[str, ...]
    risk_focus: tuple[str, ...]
    notes: str
    aliases: tuple[str, ...]
    status: str
    codegen_supported: bool


TARGET_LANGUAGES: tuple[TargetLanguageInfo, ...] = (
    TargetLanguageInfo(
        id="python-3x",
        label="Python 3",
        language="Python",
        version="3.12",
        extension=".py",
        comment_prefix="#",
        code_fence="python",
        test_framework="pytest",
        numeric_policy="Use decimal.Decimal for money and regulated calculations; never float.",
        date_policy="Use datetime.date/datetime.datetime with explicit timezone handling.",
        style_guide="PEP 8, typed public functions, small modules.",
        type_system="Type hints on public interfaces; dataclasses or Pydantic for records.",
        async_model="asyncio for I/O; external workers for CPU-heavy workloads.",
        recommended_libraries=("decimal", "datetime", "pydantic", "pytest"),
        risk_focus=("Decimal precision", "blank/null handling", "working-storage state"),
        notes="Active MVP target. All monetary math uses decimal.Decimal.",
        aliases=("Python", "python", "Python 3", "Python 3.12", "py3"),
        status="active",
        codegen_supported=True,
    ),
    TargetLanguageInfo(
        id="java-21",
        label="Java 21",
        language="Java",
        version="21",
        extension=".java",
        comment_prefix="//",
        code_fence="java",
        test_framework="JUnit 5",
        numeric_policy="Use BigDecimal for money; never double for ledgers or settlement.",
        date_policy="Use java.time; avoid Date/Calendar.",
        style_guide="Modern Java service style with explicit domain types.",
        type_system="Records/value objects for immutable data; Optional at API edges.",
        async_model="Virtual threads or structured concurrency for I/O; explicit transactions.",
        recommended_libraries=("java.math.BigDecimal", "java.time", "JUnit 5", "AssertJ"),
        risk_focus=("BigDecimal scale", "transaction boundaries", "shared mutable state"),
        notes="Experimental MVP target. Static validation uses javac when available.",
        aliases=("Java", "Java 21", "JVM"),
        status="active_experimental",
        codegen_supported=True,
    ),
    TargetLanguageInfo(
        id="csharp-dotnet",
        label="C# / .NET",
        language="C#",
        version=".NET 8",
        extension=".cs",
        comment_prefix="//",
        code_fence="csharp",
        test_framework="xUnit",
        numeric_policy="Use decimal for financial values with explicit MidpointRounding.",
        date_policy="Use DateOnly, TimeOnly, or DateTimeOffset where precision matters.",
        style_guide="Microsoft C# conventions with nullable reference types enabled.",
        type_system="Nullable reference types, records, and explicit domain identifiers.",
        async_model="async/await and Task-based APIs; avoid sync-over-async.",
        recommended_libraries=("System.Decimal", "System.DateTimeOffset", "xUnit"),
        risk_focus=("nullable handling", "decimal rounding", "async deadlocks"),
        notes="Experimental MVP target. Static validation uses dotnet build when available.",
        aliases=("C#", "CSharp", "CSharp .NET", ".NET", "dotnet", "csharp-dotnet"),
        status="active_experimental",
        codegen_supported=True,
    ),
    TargetLanguageInfo(
        id="cpp-23",
        label="C++23",
        language="C++",
        version="23",
        extension=".cpp",
        comment_prefix="//",
        code_fence="cpp",
        test_framework="GoogleTest",
        numeric_policy="Use fixed-point, decimal, or integer minor units for money.",
        date_policy="Use chrono with explicit calendar and timezone assumptions.",
        style_guide="C++ Core Guidelines: RAII, narrow interfaces, explicit ownership.",
        type_system="Value types, spans/views, smart pointers for ownership.",
        async_model="Explicit thread pools/event loops; avoid hidden shared state.",
        recommended_libraries=("<chrono>", "GoogleTest", "Catch2", "fmt"),
        risk_focus=("integer overflow", "object lifetime", "data races"),
        notes="Experimental MVP target. Static validation uses g++ or clang++.",
        aliases=("C++", "C++23", "CPP", "CPP23", "cpp-23"),
        status="active_experimental",
        codegen_supported=True,
    ),
    TargetLanguageInfo(
        id="rust-2024",
        label="Rust 2024",
        language="Rust",
        version="2024 edition",
        extension=".rs",
        comment_prefix="//",
        code_fence="rust",
        test_framework="cargo test",
        numeric_policy="Use rust_decimal or fixed-point domain types for money.",
        date_policy="Use chrono or time with explicit timezone/parsing policy.",
        style_guide="Rust API guidelines; rustfmt and clippy clean.",
        type_system="Enums for states, Result for failures, newtypes for identifiers.",
        async_model="Tokio or async runtimes for I/O; keep business rules sync/pure.",
        recommended_libraries=("rust_decimal", "chrono", "tokio", "proptest"),
        risk_focus=("ownership boundaries", "overflow policy", "error propagation"),
        notes="Experimental MVP target. Static validation uses rustc --edition 2024.",
        aliases=("Rust", "Rust 2024", "rust-2024"),
        status="active_experimental",
        codegen_supported=True,
    ),
    TargetLanguageInfo(
        id="sql-plsql",
        label="SQL / PL-SQL",
        language="SQL",
        version="PL/SQL / T-SQL",
        extension=".sql",
        comment_prefix="--",
        code_fence="sql",
        test_framework="tSQLt / utPLSQL",
        numeric_policy="Use DECIMAL/NUMERIC with explicit precision and scale.",
        date_policy="Use database-native date/timestamp types with documented session settings.",
        style_guide="Dialect-specific SQL with named transactions and reviewable procedures.",
        type_system="Constrained columns, lookup tables, checked parameters.",
        async_model="Rely on transaction isolation, locks, queues, and batch windows.",
        recommended_libraries=("utPLSQL", "tSQLt", "sqlparse"),
        risk_focus=("transaction isolation", "implicit conversion", "procedure side effects"),
        notes="Experimental MVP target. Static validation is sqlparse structural validation.",
        aliases=("SQL", "PLSQL", "PL/SQL", "T-SQL", "TSQL", "sql-plsql"),
        status="active_experimental",
        codegen_supported=True,
    ),
    TargetLanguageInfo(
        id="go-1x",
        label="Go",
        language="Go",
        version="1.22+",
        extension=".go",
        comment_prefix="//",
        code_fence="go",
        test_framework="go test",
        numeric_policy="Use integer minor units or decimal library for money; never float64.",
        date_policy="Use time.Time with explicit locations.",
        style_guide="Effective Go / gofmt; return and wrap errors explicitly.",
        type_system="Small interfaces, explicit structs, typed errors.",
        async_model="Goroutines/channels with context cancellation and guarded shared state.",
        recommended_libraries=("time", "testing", "testify", "shopspring/decimal"),
        risk_focus=("float64 money", "nil handling", "goroutine leaks"),
        notes="Experimental MVP target. Static validation uses go test -c when available.",
        aliases=("Go", "Golang", "go-1x"),
        status="active_experimental",
        codegen_supported=True,
    ),
    TargetLanguageInfo(
        id="typescript-5x",
        label="TypeScript",
        language="TypeScript",
        version="5.x",
        extension=".ts",
        comment_prefix="//",
        code_fence="typescript",
        test_framework="vitest",
        numeric_policy="Use decimal.js or bigint minor units for money; never JS number.",
        date_policy="Use typed date libraries or Temporal with explicit time zones.",
        style_guide="strict tsconfig, ESLint/Prettier, exhaustive switch handling.",
        type_system="Discriminated unions, branded types, no any.",
        async_model="async/await with typed Promises; no floating promises.",
        recommended_libraries=("decimal.js", "zod", "vitest", "date-fns"),
        risk_focus=("number money", "floating promises", "any leaks", "timezone parsing"),
        notes="Experimental MVP target. Static validation uses tsc --noEmit.",
        aliases=("TypeScript", "TS", "typescript-5x"),
        status="active_experimental",
        codegen_supported=True,
    ),
)


def _norm(value: str | None) -> str:
    value = "" if value is None else str(value)
    return " ".join(value.strip().casefold().split())


_BY_ALIAS: dict[str, TargetLanguageInfo] = {}
for _target in TARGET_LANGUAGES:
    _BY_ALIAS[_norm(_target.id)] = _target
    _BY_ALIAS[_norm(_target.language)] = _target
    _BY_ALIAS[_norm(_target.label)] = _target
    for _alias in _target.aliases:
        _BY_ALIAS[_norm(_alias)] = _target


def get_target_language(value: str | None) -> TargetLanguageInfo:
    """Resolve a target id, label, or alias to canonical backend metadata."""

    return _BY_ALIAS.get(_norm(value), _BY_ALIAS["python-3x"])


def is_supported_target_language(value: str | None) -> bool:
    """Return True when value is a known target id, label, or alias."""

    target = _BY_ALIAS.get(_norm(value))
    return bool(
        target
        and target.codegen_supported
        and target.status in {"active", "active_experimental"}
    )


def target_profile_payload(value: str | None) -> dict:
    """Return a prompt-compatible target profile payload."""

    target = get_target_language(value)
    return {
        "language": target.language,
        "version": target.version,
        "test_framework": target.test_framework,
        "notes": target.notes,
        "numeric_policy": target.numeric_policy,
        "date_policy": target.date_policy,
        "style_guide": target.style_guide,
        "type_system": target.type_system,
        "async_model": target.async_model,
        "recommended_libraries": list(target.recommended_libraries),
        "risk_focus": list(target.risk_focus),
    }


def safe_identifier(value: str, *, fallback: str = "legacy_lift_migration") -> str:
    ident = re.sub(r"[^0-9a-zA-Z_]+", "_", value).strip("_").lower()
    if not ident:
        ident = fallback
    if ident[0].isdigit():
        ident = f"{fallback}_{ident}"
    return ident


def demo_migration_code(target_value: str | None, *, chunk_id: str, business_rule: str) -> str:
    """Return deterministic sample code for demos and offline tests."""

    target = get_target_language(target_value)
    fn = safe_identifier(chunk_id)
    rule = " ".join(str(business_rule or "").split()) or "No confirmed business rule supplied."
    rule = rule.replace('"""', "'''").replace("*/", "* /")

    if target.id == "python-3x":
        return f'''from decimal import Decimal, ROUND_HALF_UP


def {fn}(
    balance: Decimal,
    annual_interest_rate: Decimal,
    days_in_period: int,
    bonus_rate: Decimal = Decimal("0"),
) -> Decimal:
    """Implement confirmed rule: {rule}"""
    account_master_table = "ACCOUNT_MASTER"
    effective_rate = annual_interest_rate + bonus_rate
    temp_rate = effective_rate / Decimal("100")
    period_factor = Decimal(days_in_period) / Decimal("365")
    interest_amount = balance * temp_rate * period_factor
    return interest_amount.quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
'''

    if target.id == "java-21":
        return f'''import java.math.BigDecimal;
import java.math.RoundingMode;

public final class LegacyLiftMigration {{
    private LegacyLiftMigration() {{}}

    /** Implements confirmed rule: {rule} */
    public static BigDecimal {fn}(BigDecimal balance, BigDecimal annualRate, int daysInPeriod) {{
        BigDecimal rate = annualRate.divide(BigDecimal.valueOf(100), 12, RoundingMode.HALF_UP);
        BigDecimal period = BigDecimal.valueOf(daysInPeriod).divide(BigDecimal.valueOf(365), 12, RoundingMode.HALF_UP);
        return balance.multiply(rate).multiply(period).setScale(2, RoundingMode.HALF_UP);
    }}
}}
'''

    if target.id == "csharp-dotnet":
        return f'''using System;

public static class LegacyLiftMigration
{{
    /// <summary>Implements confirmed rule: {rule}</summary>
    public static decimal {fn}(decimal balance, decimal annualRate, int daysInPeriod)
    {{
        var rate = annualRate / 100m;
        var period = daysInPeriod / 365m;
        return Math.Round(balance * rate * period, 2, MidpointRounding.AwayFromZero);
    }}
}}
'''

    if target.id == "cpp-23":
        return f'''// Implements confirmed rule: {rule}
long long {fn}(long long balance_cents, long long rate_basis_points, int days_in_period) {{
    return (balance_cents * rate_basis_points * days_in_period) / (10000LL * 365LL);
}}
'''

    if target.id == "rust-2024":
        return f'''/// Implements confirmed rule: {rule}
pub fn {fn}(balance_cents: i64, rate_basis_points: i64, days_in_period: i64) -> i64 {{
    (balance_cents * rate_basis_points * days_in_period) / (10_000 * 365)
}}
'''

    if target.id == "sql-plsql":
        return f'''-- Implements confirmed rule: {rule}
CREATE OR REPLACE FUNCTION {fn}(
    balance NUMERIC,
    annual_rate NUMERIC,
    days_in_period INTEGER
) RETURNS NUMERIC AS $$
BEGIN
    RETURN ROUND(balance * (annual_rate / 100) * (days_in_period::NUMERIC / 365), 2);
END;
$$ LANGUAGE plpgsql;
'''

    if target.id == "go-1x":
        go_fn = "".join(part.capitalize() for part in fn.split("_")) or "LegacyLiftMigration"
        return f'''package legacylift

// {go_fn} implements confirmed rule: {rule}
func {go_fn}(balanceCents int64, rateBasisPoints int64, daysInPeriod int64) int64 {{
    return (balanceCents * rateBasisPoints * daysInPeriod) / (10000 * 365)
}}
'''

    if target.id == "typescript-5x":
        return f'''// Implements confirmed rule: {rule}
export function {fn}(balanceCents: bigint, rateBasisPoints: bigint, daysInPeriod: bigint): bigint {{
  return (balanceCents * rateBasisPoints * daysInPeriod) / (10000n * 365n);
}}
'''

    return demo_migration_code("python-3x", chunk_id=chunk_id, business_rule=business_rule)

File: legacylift/client/lib/targetLanguages.ts
// lib/targetLanguages.ts - The catalog of languages LegacyLift can migrate INTO.
//
// Legacy migrations don't fan into a single language: core banking tends to go
// to Java, stored procedures stay in SQL/PL-SQL, analytics goes to Python, and
// so on. This catalog carries the per-language guidance the AI needs to produce
// *idiomatic* target code (not "Python with a different name") - numeric/money
// policy, date handling, typing, style, concurrency, and the test framework.
//
// Mirrors the backend MVP target catalog. Python is the active baseline; the
// other targets are explicit experimental MVP targets until sandboxed execution
// and deeper project-level dependency resolution are implemented.

export type TargetLanguageStatus = "active" | "active_experimental";

export interface TargetLanguage {
  /** Stable id, persisted in project config (never render this). */
  id: string;
  /** Short human label for dropdowns, e.g. "Java 21". */
  label: string;
  /** Canonical language name sent to the backend as `target_lang`. */
  language: string;
  version: string;
  /** File extension for assembled output, incl. the dot. */
  extension: string;
  /** Line-comment prefix for generated file headers. */
  commentPrefix: string;
  /** Product maturity shown in UI; non-Python MVP targets are experimental. */
  status: TargetLanguageStatus;
  /** True only when generation, static validation, and CI fixtures exist. */
  codegenSupported: boolean;
  testFramework: string;
  styleGuide: string;
  typeSystem: string;
  numericPolicy: string;
  datePolicy: string;
  asyncModel: string;
  recommendedLibraries: string[];
  riskFocus: string[];
  /** One-line summary shown in the prompt's target profile. */
  notes: string;
}

/** Snake-case payload matching the backend `_TargetProfileIn` shape. */
export interface TargetProfilePayload {
  language: string;
  version: string;
  test_framework: string;
  notes: string;
  numeric_policy: string;
  date_policy: string;
  style_guide: string;
  type_system: string;
  async_model: string;
  recommended_libraries: string[];
  risk_focus: string[];
}

export const TARGET_LANGUAGES: TargetLanguage[] = [
  {
    id: "python-3x",
    label: "Python 3",
    language: "Python",
    version: "3.12",
    extension: ".py",
    commentPrefix: "#",
    status: "active",
    codegenSupported: true,
    testFramework: "pytest",
    styleGuide: "PEP 8, formatted with Black; small, typed modules.",
    typeSystem:
      "Full type hints on public interfaces; dataclasses or Pydantic for structured records.",
    numericPolicy:
      "Use decimal.Decimal for money and regulated calculations - never float. Reproduce COBOL COMPUTE … ROUNDED with ROUND_HALF_UP.",
    datePolicy:
      "Use datetime.date / datetime.datetime with explicit time zones when timestamps cross system boundaries.",
    asyncModel:
      "asyncio for I/O concurrency; worker processes or external jobs for CPU-heavy work.",
    recommendedLibraries: ["decimal", "datetime", "pydantic", "pytest"],
    riskFocus: [
      "Decimal precision",
      "Implicit null/blank handling",
      "Global state from legacy working storage",
      "File encoding and fixed-width records",
    ],
    notes: "Analytics, internal tooling, fast migration. All money math uses decimal.Decimal.",
  },
  {
    id: "java-21",
    label: "Java 21",
    language: "Java",
    version: "21",
    extension: ".java",
    commentPrefix: "//",
    status: "active_experimental",
    codegenSupported: true,
    testFramework: "JUnit 5",
    styleGuide:
      "Modern Java service style - clear packages, records where appropriate, explicit domain types.",
    typeSystem:
      "Strong domain types, records for immutable values, sealed types for constrained variants, Optional at API edges.",
    numericPolicy:
      "Use BigDecimal for money and deterministic decimal arithmetic - never double for ledgers, fees, limits, or settlement. Set scale and RoundingMode explicitly.",
    datePolicy:
      "Use java.time types; keep time zones explicit at integration boundaries; avoid legacy Date/Calendar.",
    asyncModel:
      "Structured concurrency or virtual threads for high-throughput I/O; keep transaction boundaries explicit.",
    recommendedLibraries: ["java.math.BigDecimal", "java.time", "JUnit 5", "AssertJ"],
    riskFocus: [
      "BigDecimal scale and rounding",
      "Checked exception translation",
      "Transaction boundaries",
      "Thread safety / shared mutable state",
    ],
    notes:
      "Core banking, enterprise services, payments. Preserve double-entry and audit/transaction semantics.",
  },
  {
    id: "csharp-dotnet",
    label: "C# / .NET",
    language: "C#",
    version: ".NET 8",
    extension: ".cs",
    commentPrefix: "//",
    status: "active_experimental",
    codegenSupported: true,
    testFramework: "xUnit",
    styleGuide: "Microsoft C# conventions with nullable reference types enabled.",
    typeSystem:
      "Nullable reference types, records for immutable values, explicit domain types for identifiers and money.",
    numericPolicy:
      "Use decimal for financial values with an explicit MidpointRounding - never double for balances or reconciliation.",
    datePolicy:
      "Use DateOnly, TimeOnly, or DateTimeOffset where boundary precision matters.",
    asyncModel:
      "async/await for I/O and Task-based APIs; keep blocking calls out of request paths.",
    recommendedLibraries: ["System.Decimal", "System.DateTimeOffset", "xUnit", "FluentAssertions"],
    riskFocus: [
      "Nullable reference handling",
      "Decimal rounding",
      "Entity Framework transaction behavior",
      "Async deadlocks from sync-over-async",
    ],
    notes:
      "Microsoft-heavy workflows, line-of-business apps, Azure services. Model approval/audit state as first-class data.",
  },
  {
    id: "cpp-23",
    label: "C++23",
    language: "C++",
    version: "23",
    extension: ".cpp",
    commentPrefix: "//",
    status: "active_experimental",
    codegenSupported: true,
    testFramework: "GoogleTest",
    styleGuide: "C++ Core Guidelines - RAII, narrow interfaces, explicit ownership.",
    typeSystem:
      "Value types, spans/views for non-owning access, smart pointers for ownership, templates only where they reduce duplication safely.",
    numericPolicy:
      "Use fixed-point, decimal, or a vetted quant library for money - avoid binary floating-point drift in settlement-sensitive logic.",
    datePolicy:
      "Use <chrono> with explicit calendars and time-zone assumptions; document exchange calendars and cut-offs.",
    asyncModel:
      "Explicit thread pools, event loops, or low-latency messaging; avoid hidden shared state.",
    recommendedLibraries: ["<chrono>", "GoogleTest", "Catch2", "fmt"],
    riskFocus: [
      "Integer overflow",
      "Object lifetime",
      "Floating-point drift",
      "Data races and lock contention",
    ],
    notes:
      "Low-latency trading, pricing, risk engines. Separate pure pricing/risk functions from I/O adapters.",
  },
  {
    id: "rust-2024",
    label: "Rust 2024",
    language: "Rust",
    version: "2024 edition",
    extension: ".rs",
    commentPrefix: "//",
    status: "active_experimental",
    codegenSupported: true,
    testFramework: "cargo test",
    styleGuide: "Rust API guidelines; clippy and rustfmt clean.",
    typeSystem:
      "Enums for constrained states, Result for recoverable failures, newtypes for domain identifiers and units.",
    numericPolicy:
      "Use rust_decimal or a fixed-point domain type for money; keep overflow behavior explicit (checked / saturating / wrapping).",
    datePolicy:
      "Use chrono or time with explicit timezone and parsing policy; avoid implicit local-time assumptions.",
    asyncModel:
      "Async runtimes such as Tokio where I/O concurrency is required; keep business-rule code sync and pure when possible.",
    recommendedLibraries: ["rust_decimal", "chrono", "tokio", "proptest"],
    riskFocus: [
      "Ownership boundaries",
      "Error propagation",
      "Overflow policy",
      "Async runtime boundaries",
    ],
    notes:
      "Safe high-performance modernization. Model invalid states out of the domain; prefer Result over panics.",
  },
  {
    id: "sql-plsql",
    label: "SQL / PL-SQL",
    language: "SQL",
    version: "PL/SQL · T-SQL",
    extension: ".sql",
    commentPrefix: "--",
    status: "active_experimental",
    codegenSupported: true,
    testFramework: "tSQLt / utPLSQL",
    styleGuide:
      "Dialect-specific SQL style with named transactions and reviewable stored-procedure boundaries.",
    typeSystem:
      "Constrained columns, lookup tables, and checked parameters rather than unstructured strings.",
    numericPolicy:
      "Use DECIMAL / NUMERIC with explicit precision, scale, and rounding; avoid implicit casts that alter settlement values.",
    datePolicy:
      "Use database-native date/timestamp types with timezone and session settings documented per dialect.",
    asyncModel:
      "Rely on transaction isolation, locks, queues, and batch windows; keep concurrency assumptions in procedure contracts.",
    recommendedLibraries: ["utPLSQL", "tSQLt", "sqlparse"],
    riskFocus: [
      "Transaction isolation",
      "Implicit conversion",
      "Procedure side effects",
      "Audit-trail completeness",
    ],
    notes:
      "Stored procedures, reconciliation, settlement. Make transaction scopes explicit; preserve audit checkpoints.",
  },
  {
    id: "go-1x",
    label: "Go",
    language: "Go",
    version: "1.22",
    extension: ".go",
    commentPrefix: "//",
    status: "active_experimental",
    codegenSupported: true,
    testFramework: "go test",
    styleGuide: "Effective Go / gofmt; return errors explicitly and wrap with %w.",
    typeSystem:
      "Small interfaces, explicit structs, typed errors; avoid interface{}/any for domain data.",
    numericPolicy:
      "Use shopspring/decimal (or integer minor units) for money - never float64 for financial values.",
    datePolicy:
      "Use time.Time with explicit locations; avoid implicit local-time assumptions.",
    asyncModel:
      "Goroutines and channels for concurrency; guard shared state with sync primitives or context.",
    recommendedLibraries: ["shopspring/decimal", "time", "testing", "testify"],
    riskFocus: [
      "float64 used for money",
      "Nil pointer / interface handling",
      "Goroutine leaks",
      "Error wrapping and propagation",
    ],
    notes: "Modern services and tooling. Keep business rules in pure functions; isolate I/O behind interfaces.",
  },
  {
    id: "typescript-5x",
    label: "TypeScript",
    language: "TypeScript",
    version: "5.x",
    extension: ".ts",
    commentPrefix: "//",
    status: "active_experimental",
    codegenSupported: true,
    testFramework: "vitest",
    styleGuide: "ESLint + Prettier; exhaustive switch over discriminated unions.",
    typeSystem:
      "strict tsconfig; discriminated unions for states, branded types for identifiers and money, no any.",
    numericPolicy:
      "Use decimal.js (or bigint minor units) for money - never the JS number type for financial values.",
    datePolicy:
      "Use a typed date library (Temporal or date-fns) with explicit time zones; avoid ambiguous Date parsing.",
    asyncModel: "async/await with typed Promises; never leave promises floating.",
    recommendedLibraries: ["decimal.js", "zod", "vitest", "date-fns"],
    riskFocus: [
      "number type used for money",
      "Floating promises",
      "any leaks",
      "Timezone-ambiguous Date parsing",
    ],
    notes: "Web/service modernization. Keep pure business rules separate from framework/IO code.",
  },
];

export const DEFAULT_TARGET_ID = "python-3x";

const _BY_ID = new Map(TARGET_LANGUAGES.map((t) => [t.id, t]));

/** Resolve a catalog entry by id, falling back to the default (Python). */
export function getTargetLanguage(id: string | undefined | null): TargetLanguage {
  return (id && _BY_ID.get(id)) || _BY_ID.get(DEFAULT_TARGET_ID)!;
}

/** Build the snake_case profile payload the /llm/* prompts consume. */
export function toProfileCtx(t: TargetLanguage): TargetProfilePayload {
  return {
    language: t.language,
    version: t.version,
    test_framework: t.testFramework,
    notes: t.notes,
    numeric_policy: t.numericPolicy,
    date_policy: t.datePolicy,
    style_guide: t.styleGuide,
    type_system: t.typeSystem,
    async_model: t.asyncModel,
    recommended_libraries: t.recommendedLibraries,
    risk_focus: t.riskFocus,
  };
}
